### ISSUE TITLE
model map is not thread  safe ,and my lock didn't effect

### ISSUE THREAD

**Issue by tokingHong** [casbin/casbin#191]
I used to load policy in every request. But sometime i could get a same policy twice.
I kenw the map is not thread safe. I looked source code, and I guessed it would be ok when I lock every load policy .But it did not effect....
![image](https://user-images.githubusercontent.com/39825820/50482979-76d0b300-0a24-11e9-94fb-b93f1c6d309b.png)
![image](https://user-images.githubusercontent.com/39825820/50482996-8cde7380-0a24-11e9-8252-41e9a2bab39c.png)
  I knew it's not reasonable to use casbin in this way(and I have changed).I just need some guys can tell me why my lock not effect?
![image](https://user-images.githubusercontent.com/39825820/50483124-3faed180-0a25-11e9-8e27-a7640cf6670d.png)

**Comment by hsluoyz** [comment]
Can you give a minimized working example?

**Comment by hsluoyz** [comment]
This is not a complete runnable example.

**Comment by kenjones-cisco** [comment]
The use case where we noticed the same is when there is a single instance of `Enforcer`, and then using `LoadFilteredPolicy` and `Enforce`, as the number of concurrent requests increase the results from `Enforce` become unexpected.

Each thread is effectively emptying the Model and then re-loading with only the policies that apply to the specified context, and then calling `Enforce`. As such we would see times where no policies were available when `Enforce` was called because another thread had just started the load step.

To resolve the issue, I have created a Pool of Enforcer, so that each thread retrieves an Enforcer from the pool and loads required policies and calls `Enforce`, and then returns the Enforcer to the pool for use by another thread. The pool is a buffered channel that will grow as needed and shrink if the Enforcer fails with an error.
